src/goto-instrument/contracts/assigns.cpp

```diff
@@ -56,25 +55,64 @@ assigns_clause_targett::~assigns_clause_targett()
 std::vector<symbol_exprt> assigns_clause_targett::temporary_declarations() const
 {
   std::vector<symbol_exprt> result;
-  result.push_back(local_target);
+  result.push_back(target);
   return result;
 }
 
-exprt assigns_clause_targett::alias_expression(const exprt &ptr)
+exprt assigns_clause_targett::alias_expression(const exprt &lhs)
 {
-  return same_object(ptr, local_target);
+  exprt::operandst condition;
+  exprt lhs_ptr = (lhs.id() == ID_address_of) ? to_address_of_expr(lhs).object()
+                                              : pointer_for(lhs);
+
+  // __CPROVER_same_object(lhs, target)
+  condition.push_back(same_object(lhs_ptr, target));
+
+  // If assigns target was a dereference, comparing objects is enough
+  if(target_id == ID_dereference)
+  {
+    return conjunction(condition);
+  }
+
+  if(target_id == ID_dereference)
+  {
+    return conjunction(condition);
+  }
```

Comment:
Remove the 4 lines above? They look repeated

Reply:
Rebase problem. Removed it.